File: C++/labs/lab2/avlChecker.cpp
// Задание 4 | Вариант 3

#include "avlChecker.h"

#include <iostream>
#include <string>
#include <vector>

#include "binaryTree.h"

void runAvlChecker(int argc, char* argv[]) {
    // Создаем экземпляр нашего бинарного дерева.
    BinaryTree tree;

    std::cout << "Введите последовательность целых чисел, оканчивающуюся нулем:" << std::endl;

    int number;
    // Читаем числа из стандартного ввода, пока не встретим 0.
    while (std::cin >> number && number != 0) {
        // Вставляем каждое число в дерево, предварительно сконвертировав в строку.
        tree.insert(std::to_string(number));
    }

    // Вызываем наш новый метод для проверки сбалансированности.
    if (tree.isAVLBalanced()) {
        std::cout << "YES" << std::endl;
    } else {
        std::cout << "NO" << std::endl;
    }
}

File: Go/labs/lab2/avlChecker.go
package main

import (
	"fmt"
	"math"
)

type TreeNode struct {
	Value string
	Left  *TreeNode
	Right *TreeNode
}

type BinaryTree struct {
	Root *TreeNode
}

func (t *BinaryTree) insertRec(node *TreeNode, value string) *TreeNode {
	if node == nil {
		return &TreeNode{Value: value}
	}
	if value < node.Value {
		node.Left = t.insertRec(node.Left, value)
	} else if value > node.Value {
		node.Right = t.insertRec(node.Right, value)
	}
	return node
}

func (t *BinaryTree) Insert(value string) {
	t.Root = t.insertRec(t.Root, value)
}

func height(node *TreeNode) int {
	if node == nil {
		return 0
	}
	return 1 + int(math.Max(float64(height(node.Left)), float64(height(node.Right))))
}

func (t *BinaryTree) isAVLBalancedRec(node *TreeNode) bool {
	if node == nil {
		return true
	}
	leftHeight := height(node.Left)
	rightHeight := height(node.Right)

	if math.Abs(float64(leftHeight-rightHeight)) > 1 {
		return false
	}
	return t.isAVLBalancedRec(node.Left) && t.isAVLBalancedRec(node.Right)
}

func (t *BinaryTree) IsAVLBalanced() bool {
	return t.isAVLBalancedRec(t.Root)
}

func runAvlChecker() {
	tree := &BinaryTree{}
	fmt.Println("Введите последовательность целых чисел, оканчивающуюся нулем:")
	for {
		var number int
		_, err := fmt.Scan(&number)
		if err != nil || number == 0 {
			break
		}
		tree.Insert(fmt.Sprint(number))
	}

	if tree.IsAVLBalanced() {
		fmt.Println("YES")
	} else {
		fmt.Println("NO")
	}
}